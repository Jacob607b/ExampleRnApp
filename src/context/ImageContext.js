import { createContext } from "react";

export const ImageContext = createContext({ Images: [null] });

export default ImageContext;
