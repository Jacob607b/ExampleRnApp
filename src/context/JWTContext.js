import { createContext } from "react";

export const JWTContext = createContext({ Jwt: undefined });

export default JWTContext;
