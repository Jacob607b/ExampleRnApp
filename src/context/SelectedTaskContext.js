import { createContext } from "react";

export const SelectedTaskContext = createContext({selectedTask : []});

export default SelectedTaskContext;