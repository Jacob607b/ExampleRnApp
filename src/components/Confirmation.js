import {
  ScrollView,
  Text,
  StyleSheet,
  TouchableOpacity,
  View,
  BackHandler,
  Alert,
} from "react-native";
import React, { useContext, useState, useEffect } from "react";
import Image from "react-native-image-progress";
import { ReadContext } from "../context/ReadContext";
import { IDContext } from "../context/IDContext";
import { ImageContext } from "../context/ImageContext";
import { SelectedTaskContext } from "../context/SelectedTaskContext";
import { getAuth } from "firebase/auth";
import { collection, addDoc } from "firebase/firestore";
import { Button, ActivityIndicator } from "react-native-paper";
import uuid from "react-native-uuid";
import * as Location from "expo-location";
import { getUnixTime } from "date-fns";
import { getFirestore } from "firebase/firestore";
import { ref, getStorage } from "firebase/storage";
// DO NOT REMOVE storage import
import { storage } from "../firebase";

const Confirmation = React.memo(({ navigation }) => {
  const { readData } = useContext(ReadContext);
  const { ID } = useContext(IDContext);
  const { Images } = useContext(ImageContext);
  const { selectedTask } = useContext(SelectedTaskContext);
  const storage = getStorage();

  // Hook to set location, error message if found, and isLoaded to lazy load map comonent if needed
  const [local, setLocation] = useState([]);
  const [errorMsg, setErrorMsg] = useState(null);

  const [submitIndicator, setSubmitIndicator] = useState(false);
  //Device location is set if User allows it
  const getLocation = async () => {
    let { status } = await Location.requestForegroundPermissionsAsync();
    if (status !== "granted") {
      setErrorMsg("Permission to access location was denied");
      alert(errorMsg);
    }
    let location = await Location.getCurrentPositionAsync({});
    setLocation(location);
  };
  useEffect(() => {
    getLocation();
  }, []);

  const auth = getAuth();
  const user = auth.currentUser;
  const date = new Date();
  const year = date.getFullYear();
  const month = date.getMonth() + 1;
  const dt = date.getDate();
  const UnixTime = getUnixTime(date);
  // Array of month abbreviations
  const monthAbbreviations = [
    "Jan",
    "Feb",
    "Mar",
    "Apr",
    "May",
    "Jun",
    "Jul",
    "Aug",
    "Sep",
    "Oct",
    "Nov",
    "Dec",
  ];
  // Determine the cycle based on the date
  const cycle = `${monthAbbreviations[month - 1]}${year.toString().slice(2)}`;
  const handleSubmit = async () => {
    setSubmitIndicator(true);
    let FireStoreData;
    if (selectedTask != null) {
      FireStoreData = {
        Data: { addedField: readData.addedField },
        Unit: readData.address,
        Images,
        local,
        Date: `${year}-${month < 10 ? "0" : ""}${month}-${
          dt < 10 ? "0" : ""
        }${dt}`,
        ID: ID,
        UnixTime,
        Cycle: selectedTask.Cycle,
      };
    } else {
      FireStoreData = {
        Data: { addedField: readData.addedField },
        Unit: readData.address,
        Images,
        local,
        Date: `${year}-${month < 10 ? "0" : ""}${month}-${
          dt < 10 ? "0" : ""
        }${dt}`,
        ID: ID,
        UnixTime,
        Cycle: cycle,
      };
    }

    const docRef = collection(getFirestore(), `Reads/${ID}/reads`);
    await addDoc(docRef, FireStoreData);
    try {
      const token = user && (await user.getIdToken());
      await fetch(
        "http://10.0.2.2:5001/vcs-reads-dev2/us-central1/app/updateTask",
        {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
            Authorization: `Bearer ${token}`,
          },
          body: JSON.stringify({
            ID: ID,
            taskString: selectedTask.Description,
            completed: true,
            Unit: selectedTask.Unit,
            Cycle: selectedTask.Cycle,
          }),
        }
      );
    } catch (error) {
      console.log(error);
    }
    setSubmitIndicator(false);
    navigateToWelcome();
  };

  const cancelSubmission = async () => {
    const token = user && (await user.getIdToken());
    await Images.ImageURL.map(async (image) => {
      const fileName = ref(storage, image);
      const name = fileName.name;
      const requestOptions = {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${token}`,
        },
        body: JSON.stringify({ ID: ID, URL: name }),
      };
      await fetch(
        "http://10.0.2.2:5001/vcs-reads-dev2/us-central1/app/delete",
        requestOptions
      );
    });
    await navigation.navigate("Welcome");
  };

  const navigateToWelcome = async () => {
    await navigation.navigate("Welcome");
  };
  const alertFunciton = async () => {
    Alert.alert("Hold on!", "Are you sure you want to go back?", [
      {
        text: "Cancel",
        onPress: () => null,
        style: "cancel",
      },
      {
        text: "YES",
        onPress: async () => {
          await cancelSubmission(), await navigateToWelcome();
        },
      },
    ]);
  };
  const onBackPress = async () => {
    alertFunciton();
  };
  BackHandler.addEventListener("hardwareBackPress", onBackPress);

  const showIndicator = () => {
    return <ActivityIndicator size="large" color={"#8fc33b"} />;
  };
  return (
    <>
      <ScrollView>
        <View style={styles.container}>
          <Text style={styles.largFont}>UNIT: {readData.address}</Text>
          <>
            {readData.addedField.map((added) => {
              return (
                <>
                  <View style={styles.inputContainer} key={uuid.v4()}>
                    <Text key={uuid.v4()} style={styles.largFont}>
                      Type: {added.type}
                    </Text>
                  </View>

                  <Text style={styles.largFont} key={uuid.v4()}>
                    Read: {added.read}
                  </Text>
                </>
              );
            })}
          </>

          <ScrollView
            horizontal
            showsHorizontalScrollIndicator={true}
            contentContainerStyle={styles.scrollView}
          >
            {Images.ImageURL.map((url) => {
              return (
                <View style={styles.imageContainer} key={uuid.v4()}>
                  <Image
                    key={url}
                    source={{ uri: url }}
                    style={styles.image}
                    indicator={showIndicator}
                  />
                </View>
              );
            })}
          </ScrollView>
          {submitIndicator ? (
            <ActivityIndicator color="#8fc33b" />
          ) : (
            <TouchableOpacity
              onPress={handleSubmit}
              testID="Confirm-Submission"
            >
              <Text>
                <Button mode="elevated" testID="Submit">
                  Submit
                </Button>
              </Text>
            </TouchableOpacity>
          )}

          <TouchableOpacity onPress={cancelSubmission} testID="cancel">
            <Button mode="elevated" style={styles.cancelButton}>
              Cancel
            </Button>
          </TouchableOpacity>
        </View>
      </ScrollView>
    </>
  );
});
// Style-sheet object
const styles = StyleSheet.create({
  container: {
    flex: 1,
    width: "100%",
    backgroundColor: "#fff",
    alignItems: "center",
    justifyContent: "center",
  },
  inputContainer: {
    flexDirection: "column",
    padding: 10,
    justifyContent: "space-between",
    alignItems: "center",
  },
  txt: {
    color: "white",
  },
  largFont: {
    fontSize: 20,
  },
  scrollView: {
    flexGrow: 1,
    justifyContent: "center",
  },
  imageContainer: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
    margin: 10,
  },
  image: {
    flex: 1,
    height: 500,
    width: 200,
  },
  cancelButton: {
    marginTop: 10,
    marginBottom: 10,
  },
});
export default Confirmation;
