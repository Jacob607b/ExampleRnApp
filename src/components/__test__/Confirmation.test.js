jest.useFakeTimers();
import React from "react";
import { ReadContext } from "../../context/ReadContext";
import { ImageContext } from "../../context/ImageContext";
import { SelectedTaskContext } from "../../context/SelectedTaskContext";
import {
  cleanup,
  fireEvent,
  render,
  screen,
  act,
} from "@testing-library/react-native";
import Confirmation from "../Confirmation";
import * as Location from "expo-location";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { BackHandler } from "react-native";

jest.mock("react-native/Libraries/Utilities/BackHandler", () => {
  return jest.requireActual(
    "react-native/Libraries/Utilities/__mocks__/BackHandler.js"
  );
});
const queryClient = new QueryClient();
let useEffect;
const mockUseEffect = () => {
  useEffect.mockImplementation((f) => f());
};
beforeEach(() => {
  useEffect = jest.spyOn(React, "useEffect");
  mockUseEffect();
  mockUseEffect();
});

afterEach(cleanup);
// Location Permission spyOn
const locationPermSpy = jest.spyOn(
  Location,
  "requestForegroundPermissionsAsync"
);

const locationSpy = jest.spyOn(Location, "getCurrentPositionAsync");

test("Renders, snapshot, state updates", async () => {
  global.fetch = jest.fn().mockResolvedValueOnce({
    status: 200,
  });
  // Location Permission status is mocked
  locationPermSpy.mockResolvedValueOnce({ status: "granted" });

  // readData is needed for Confirmation component to render, maps through the value of Context in component
  const readData = {
    address: "Unit 1",
    addedField: [{ field: "Electric", read: "100" }],
  };
  const Images = { ImageURL: ["image.source"] };
  const navigate = jest.fn();
  const date = jest.fn();
  dt = jest.fn();
  const requestOptions = jest.fn();
  const setSelectedTask = jest.fn();

  const task = { selectedTask: ["March24"] };

  render(
    <QueryClientProvider client={queryClient}>
      <SelectedTaskContext.Provider value={{ task, setSelectedTask }}>
        <ReadContext.Provider value={{ readData, requestOptions }}>
          <ImageContext.Provider value={{ Images }}>
            <Confirmation
              navigation={{ navigate }}
              date={{ date }}
              dt={dt}
              location={locationSpy}
            />
          </ImageContext.Provider>
        </ReadContext.Provider>
      </SelectedTaskContext.Provider>
    </QueryClientProvider>
  );
  expect(locationPermSpy).toHaveBeenCalled();

  expect(screen).toMatchSnapshot();

  const confirm = screen.getByTestId("Confirm-Submission");
  expect(confirm).toBeOnTheScreen();

  fireEvent.press(confirm);

  const cancel = screen.getByTestId("cancel");
  expect(cancel).toBeOnTheScreen();
  fireEvent.press(cancel);
  BackHandler.mockPressBack();
});
test("Location permission not granted", async () => {
  global.fetch = jest.fn().mockResolvedValueOnce({
    status: 200,
  });
  await act(() => {
    // Location Permission status is mocked
    locationPermSpy.mockResolvedValueOnce({ status: "NOPE" });
  });

  // readData is needed for Confirmation component to render, maps through the value of Context in component
  const readData = {
    address: "Unit 1",
    addedField: [{ field: "Electric", read: "100" }],
  };
  const Images = { ImageURL: ["image.source"] };
  const navigate = jest.fn();
  const date = jest.fn();
  dt = jest.fn();
  const requestOptions = jest.fn();
  const setSelectedTask = jest.fn();

  const task = { selectedTask: ["March24"] };

  render(
    <QueryClientProvider client={queryClient}>
      <SelectedTaskContext.Provider value={{ task, setSelectedTask }}>
        <ReadContext.Provider value={{ readData, requestOptions }}>
          <ImageContext.Provider value={{ Images }}>
            <Confirmation
              navigation={{ navigate }}
              date={{ date }}
              dt={dt}
              location={locationSpy}
            />
          </ImageContext.Provider>
        </ReadContext.Provider>
      </SelectedTaskContext.Provider>
    </QueryClientProvider>
  );
  expect(locationPermSpy).toHaveBeenCalled();
});
