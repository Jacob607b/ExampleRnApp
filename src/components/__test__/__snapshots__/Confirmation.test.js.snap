// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders, snapshot, state updates 1`] = `
<RCTScrollView>
  <View>
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "flex": 1,
          "justifyContent": "center",
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 20,
          }
        }
      >
        UNIT: 
        Unit 1
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "column",
            "justifyContent": "space-between",
            "padding": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 20,
            }
          }
        >
          Type: 
        </Text>
      </View>
      <Text
        style={
          Object {
            "fontSize": 20,
          }
        }
      >
        Read: 
        100
      </Text>
      <RCTScrollView
        contentContainerStyle={
          Object {
            "flexGrow": 1,
            "justifyContent": "center",
          }
        }
        horizontal={true}
        showsHorizontalScrollIndicator={true}
      >
        <View>
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "justifyContent": "center",
                "margin": 10,
              }
            }
          >
            <View
              style={
                Object {
                  "flex": 1,
                  "height": 500,
                  "width": 200,
                }
              }
            >
              <Image
                onError={[Function]}
                onLoad={[Function]}
                onLoadEnd={[Function]}
                onLoadStart={[Function]}
                onProgress={[Function]}
                source={
                  Object {
                    "uri": "image.source",
                  }
                }
                style={
                  Array [
                    Object {
                      "bottom": 0,
                      "left": 0,
                      "position": "absolute",
                      "right": 0,
                      "top": 0,
                    },
                    undefined,
                  ]
                }
              />
            </View>
          </View>
        </View>
      </RCTScrollView>
      <View
        accessibilityState={
          Object {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          Object {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="Confirm-Submission"
      >
        <Text>
          <View
            collapsable={false}
            style={
              Object {
                "backgroundColor": "rgb(247, 243, 249)",
                "borderRadius": 20,
                "shadowColor": "#000",
                "shadowOffset": Object {
                  "height": 1,
                  "width": 0,
                },
                "shadowOpacity": 0.15,
                "shadowRadius": 3,
              }
            }
            testID="Submit-container-outer-layer"
          >
            <View
              collapsable={false}
              style={
                Object {
                  "backgroundColor": "rgb(247, 243, 249)",
                  "borderColor": "transparent",
                  "borderRadius": 20,
                  "borderStyle": "solid",
                  "borderWidth": 0,
                  "flex": undefined,
                  "minWidth": 64,
                  "shadowColor": "#000",
                  "shadowOffset": Object {
                    "height": 1,
                    "width": 0,
                  },
                  "shadowOpacity": 0.3,
                  "shadowRadius": 1,
                }
              }
              testID="Submit-container"
            >
              <View
                accessibilityRole="button"
                accessibilityState={
                  Object {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": true,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  Object {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Array [
                    Object {
                      "overflow": "hidden",
                    },
                    Object {
                      "borderRadius": 20,
                    },
                  ]
                }
                testID="Submit"
              >
                <View
                  style={
                    Array [
                      Object {
                        "alignItems": "center",
                        "flexDirection": "row",
                        "justifyContent": "center",
                      },
                      undefined,
                    ]
                  }
                >
                  <Text
                    numberOfLines={1}
                    selectable={false}
                    style={
                      Array [
                        Object {
                          "textAlign": "left",
                        },
                        Object {
                          "color": "rgba(28, 27, 31, 1)",
                          "writingDirection": "ltr",
                        },
                        Array [
                          Object {
                            "fontFamily": "System",
                            "fontSize": 14,
                            "fontWeight": "500",
                            "letterSpacing": 0.1,
                            "lineHeight": 20,
                          },
                          Array [
                            Object {
                              "marginHorizontal": 16,
                              "marginVertical": 9,
                              "textAlign": "center",
                            },
                            false,
                            Object {
                              "marginHorizontal": 24,
                              "marginVertical": 10,
                            },
                            undefined,
                            false,
                            Object {
                              "color": "rgba(103, 80, 164, 1)",
                              "fontFamily": "System",
                              "fontSize": 14,
                              "fontWeight": "500",
                              "letterSpacing": 0.1,
                              "lineHeight": 20,
                            },
                            undefined,
                          ],
                        ],
                      ]
                    }
                    testID="Submit-text"
                  >
                    Submit
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </Text>
      </View>
      <View
        accessibilityState={
          Object {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          Object {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="cancel"
      >
        <View
          collapsable={false}
          style={
            Object {
              "backgroundColor": "rgb(247, 243, 249)",
              "borderRadius": 20,
              "marginBottom": 10,
              "marginTop": 10,
              "shadowColor": "#000",
              "shadowOffset": Object {
                "height": 1,
                "width": 0,
              },
              "shadowOpacity": 0.15,
              "shadowRadius": 3,
            }
          }
          testID="button-container-outer-layer"
        >
          <View
            collapsable={false}
            style={
              Object {
                "backgroundColor": "rgb(247, 243, 249)",
                "borderColor": "transparent",
                "borderRadius": 20,
                "borderStyle": "solid",
                "borderWidth": 0,
                "flex": undefined,
                "minWidth": 64,
                "shadowColor": "#000",
                "shadowOffset": Object {
                  "height": 1,
                  "width": 0,
                },
                "shadowOpacity": 0.3,
                "shadowRadius": 1,
              }
            }
            testID="button-container"
          >
            <View
              accessibilityRole="button"
              accessibilityState={
                Object {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": true,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                Object {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Array [
                  Object {
                    "overflow": "hidden",
                  },
                  Object {
                    "borderRadius": 20,
                  },
                ]
              }
              testID="button"
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "center",
                    },
                    undefined,
                  ]
                }
              >
                <Text
                  numberOfLines={1}
                  selectable={false}
                  style={
                    Array [
                      Object {
                        "textAlign": "left",
                      },
                      Object {
                        "color": "rgba(28, 27, 31, 1)",
                        "writingDirection": "ltr",
                      },
                      Array [
                        Object {
                          "fontFamily": "System",
                          "fontSize": 14,
                          "fontWeight": "500",
                          "letterSpacing": 0.1,
                          "lineHeight": 20,
                        },
                        Array [
                          Object {
                            "marginHorizontal": 16,
                            "marginVertical": 9,
                            "textAlign": "center",
                          },
                          false,
                          Object {
                            "marginHorizontal": 24,
                            "marginVertical": 10,
                          },
                          undefined,
                          false,
                          Object {
                            "color": "rgba(103, 80, 164, 1)",
                            "fontFamily": "System",
                            "fontSize": 14,
                            "fontWeight": "500",
                            "letterSpacing": 0.1,
                            "lineHeight": 20,
                          },
                          undefined,
                        ],
                      ],
                    ]
                  }
                  testID="button-text"
                >
                  Cancel
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
